BDArmory/Control/BDModuleOrbitalAI.cs

```diff
@@ -0,0 +1,1041 @@
+                }
+            }
+            else
+            {
+                // Idle
+
+                if (hasWeapons)
+                    currentStatus = "Idle";
+                else
+                    currentStatus = "Idle (Unarmed)";
+
+                fc.throttle = 0;
+                fc.attitude = Vector3.zero;
+
+                yield return new WaitForSeconds(updateInterval);
+            }
+        }
+
+        #endregion Actual AI Pilot
+
+        #region Utility Functions
+
+        private bool CheckWithdraw()
+        {
+            var nearest = BDATargetManager.GetClosestTarget(weaponManager);
+            if (nearest == null) return false;
+
+            return Mathf.Abs(RelVel(vessel, nearest.Vessel).magnitude) < 200;
+        }
+
+        private bool CheckOrbitUnsafe()
+        {
+            Orbit o = vessel.orbit;
+            CelestialBody body = o.referenceBody;
+            double maxTerrainHeight = 200;
+            if (body.pqsController)
+            {
+                PQS pqs = body.pqsController;
+                maxTerrainHeight = pqs.radiusMax - pqs.radius;
+            }
+            minSafeAltitude = Math.Max(maxTerrainHeight, body.atmosphereDepth);
+
+            return (o.PeA < minSafeAltitude && o.timeToPe < o.timeToAp) || o.ApA < minSafeAltitude;
+        }
+
+        private bool UnderTimeLimit(float timeLimit = 0)
+        {
+            if (timeLimit == 0)
+                timeLimit = updateInterval;
+
+            return Time.time - lastUpdate < timeLimit;
+        }
+
+        private bool NearIntercept(Vector3 relVel, float minRange)
+        {
+            float timeToKillVelocity = relVel.magnitude / maxAcceleration;
+
+            float rotDistance = Vector3.Angle(vessel.ReferenceTransform.up, relVel.normalized * -1) * Mathf.Deg2Rad;
+            float timeToRotate = SolveTime(rotDistance * 0.75f, maxAngularAcceleration.magnitude) / 0.75f;
+
+            Vector3 toTarget = FromTo(vessel, targetVessel);
+            Vector3 toClosestApproach = ToClosestApproach(toTarget, relVel, minRange);
+
+            // Return false if we aren't headed towards the target.
+            float velToClosestApproach = Vector3.Dot(relVel, toTarget.normalized);
+            if (velToClosestApproach < 10)
+                return false;
+
+            float timeToClosestApproach = AIUtils.TimeToCPA(toClosestApproach, relVel * -1, Vector3.zero, 9999);
+            if (timeToClosestApproach == 0)
+                return false;
+
+            nearInterceptBurnTime = timeToKillVelocity + timeToRotate;
+            nearInterceptApproachTime = timeToClosestApproach;
+
+            return timeToClosestApproach < (timeToKillVelocity + timeToRotate);
+        }
+
+        private bool CanInterceptShip(Vessel target)
+        {
+            bool canIntercept = false;
+
+            // Is it worth us chasing a withdrawing ship?
+            BDModuleOrbitalAI targetAI = target.FindPartModuleImplementing<BDModuleOrbitalAI>();
```

Comment:
This should also use `VesselModuleRegistry`.

Reply:
I meant that this should use
```csharp
BDModuleOrbitalAI targetAI = VesselModuleRegistry.GetModule<BDModuleOrbitalAI>(target);
```
now and I can later add the specialised registries to avoid the boxing/unboxing overhead.